File: en/codes/c/chapter_backtracking/preorder_traversal_iii_template.c
/**
 * File: preorder_traversal_iii_template.c
 * Created Time: 2023-06-04
 * Author: Gonglja (glj0@outlook.com)
 */

#include "../utils/common.h"

// Assume path sum result length does not exceed 100
#define MAX_SIZE 100
#define MAX_RES_SIZE 100

TreeNode *path[MAX_SIZE];
TreeNode *res[MAX_RES_SIZE][MAX_SIZE];
int pathSize = 0, resSize = 0;

/* Determine if the current state is a solution */
bool isSolution(void) {
    return pathSize > 0 && path[pathSize - 1]->val == 7;
}

/* Record solution */
void recordSolution(void) {
    for (int i = 0; i < pathSize; i++) {
        res[resSize][i] = path[i];
    }
    resSize++;
}

/* Determine if the choice is legal under the current state */
bool isValid(TreeNode *choice) {
    return choice != NULL && choice->val != 3;
}

/* Update state */
void makeChoice(TreeNode *choice) {
    path[pathSize++] = choice;
}

/* Restore state */
void undoChoice(void) {
    pathSize--;
}

/* Backtracking algorithm: Example three */
void backtrack(TreeNode *choices[2]) {
    // Check if it's a solution
    if (isSolution()) {
        // Record solution
        recordSolution();
    }
    // Traverse all choices
    for (int i = 0; i < 2; i++) {
        TreeNode *choice = choices[i];
        // Pruning: check if the choice is legal
        if (isValid(choice)) {
            // Attempt: make a choice, update the state
            makeChoice(choice);
            // Proceed to the next round of selection
            TreeNode *nextChoices[2] = {choice->left, choice->right};
            backtrack(nextChoices);
            // Retract: undo the choice, restore to the previous state
            undoChoice();
        }
    }
}

/* Driver Code */
int main() {
    int arr[] = {1, 7, 3, 4, 5, 6, 7};
    TreeNode *root = arrayToTree(arr, sizeof(arr) / sizeof(arr[0]));
    printf("\nInitialize binary tree\n");
    printTree(root);

    // Backtracking algorithm
    TreeNode *choices[2] = {root, NULL};
    backtrack(choices);

    printf("\nOutput all paths from root to node 7, excluding paths containing value 3\n");
    for (int i = 0; i < resSize; ++i) {
        int *vals = malloc(MAX_SIZE * sizeof(int));
        int size = 0;
        for (int j = 0; res[i][j] != NULL; ++j) {
            vals[size++] = res[i][j]->val;
        }
        printArray(vals, size);
        free(vals);
    }

    // Free memory
    freeMemoryTree(root);
    return 0;
}

File: en/codes/dart/chapter_backtracking/preorder_traversal_iii_template.dart
/**
 * File: preorder_traversal_iii_template.dart
 * Created Time: 2023-08-10
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

import '../utils/print_util.dart';
import '../utils/tree_node.dart';

/* Determine if the current state is a solution */
bool isSolution(List<TreeNode> state) {
  return state.isNotEmpty && state.last.val == 7;
}

/* Record solution */
void recordSolution(List<TreeNode> state, List<List<TreeNode>> res) {
  res.add(List.from(state));
}

/* Determine if the choice is legal under the current state */
bool isValid(List<TreeNode> state, TreeNode? choice) {
  return choice != null && choice.val != 3;
}

/* Update state */
void makeChoice(List<TreeNode> state, TreeNode? choice) {
  state.add(choice!);
}

/* Restore state */
void undoChoice(List<TreeNode> state, TreeNode? choice) {
  state.removeLast();
}

/* Backtracking algorithm: Example three */
void backtrack(
  List<TreeNode> state,
  List<TreeNode?> choices,
  List<List<TreeNode>> res,
) {
  // Check if it's a solution
  if (isSolution(state)) {
    // Record solution
    recordSolution(state, res);
  }
  // Traverse all choices
  for (TreeNode? choice in choices) {
    // Pruning: check if the choice is legal
    if (isValid(state, choice)) {
      // Attempt: make a choice, update the state
      makeChoice(state, choice);
      // Proceed to the next round of selection
      backtrack(state, [choice!.left, choice.right], res);
      // Retract: undo the choice, restore to the previous state
      undoChoice(state, choice);
    }
  }
}

/* Driver Code */
void main() {
  TreeNode? root = listToTree([1, 7, 3, 4, 5, 6, 7]);
  print("\nInitialize binary tree");
  printTree(root);

  // Backtracking algorithm
  List<List<TreeNode>> res = [];
  backtrack([], [root!], res);
  print("\nOutput all root-to-node 7 paths, requiring paths not to include nodes with value 3");
  for (List<TreeNode> path in res) {
    print(List.from(path.map((e) => e.val)));
  }
}
